src/cls/rgw/cls_rgw.cc

```diff
@@ -3472,6 +3472,273 @@ static int rgw_cls_lc_get_head(cls_method_context_t hctx, bufferlist *in,  buffe
+  cls_rgw_reshard_entry entry;
+  for (it = vals.begin(); it != vals.end(); ++it) {
+    iter = it->second.begin();
+    try {
+    ::decode(entry, iter);
+    } catch (buffer::error& err) {
+    CLS_LOG(1, "ERROR: rgw_cls_rehard_list(): failed to decode entry\n");
+    return -EIO;
+   }
+    op_ret.entries.push_back(entry);
+  }
+  op_ret.is_truncated = op.max && (op_ret.entries.size() >= op.max);
+  ::encode(op_ret, *out);
+  return 0;
+}
+
+
+static int rgw_reshard_get_head(cls_method_context_t hctx, bufferlist *in,  bufferlist *out)
+{
+  bufferlist bl;
+  int ret = cls_cxx_map_read_header(hctx, &bl);
+  if (ret < 0)
+    return ret;
+  cls_rgw_reshard_entry  entry;
+  if (bl.length() != 0) {
+    bufferlist::iterator iter = bl.begin();
+    try {
+      ::decode(entry, iter);
+    } catch (buffer::error& err) {
+      CLS_LOG(0, "ERROR: rgw_reshard_get_head(): failed to decode entry %s\n",err.what());
+      return -EINVAL;
+    }
+  }
+
+  cls_rgw_reshard_get_head_ret op_ret;
+  op_ret.entry = entry;
+  ::encode(op_ret, *out);
+  return 0;
+}
+
+static int rgw_reshard_get(cls_method_context_t hctx, bufferlist *in,  bufferlist *out)
+{
+  bufferlist::iterator in_iter = in->begin();
+
+  cls_rgw_reshard_get_op op;
+  try {
+    ::decode(op, in_iter);
+  } catch (buffer::error& err) {
+    CLS_LOG(1, "ERROR: rgw_reshard_get: failed to decode entry\n");
+    return -EINVAL;
+  }
+
+  bufferlist bl;
+  string key = op.entry.bucket_name + "." + op.entry.bucket_id;
+  int ret = cls_cxx_map_get_val(hctx, key, &bl);
+  if (ret < 0)
+    return ret;
+  cls_rgw_reshard_entry  entry;
+  if (bl.length() != 0) {
+    bufferlist::iterator iter = bl.begin();
+    try {
+      ::decode(entry, iter);
+    } catch (buffer::error& err) {
+      CLS_LOG(0, "ERROR: rgw_cls_reshard_get_head(): failed to decode entry %s\n",err.what());
+      return -EINVAL;
+    }
+  }
+
+  cls_rgw_reshard_get_ret op_ret;
+  op_ret.entry = entry;
+  ::encode(op_ret, *out);
+  return 0;
+}
+
+static int rgw_reshard_remove(cls_method_context_t hctx, bufferlist *in, bufferlist *out)
+{
+  bufferlist::iterator in_iter = in->begin();
+
+  cls_rgw_reshard_remove_op op;
+  try {
+    ::decode(op, in_iter);
+  } catch (buffer::error& err) {
+    CLS_LOG(1, "ERROR: rgw_cls_rehard_remove: failed to decode entry\n");
+    return -EINVAL;
+  }
+
+  string key = op.bucket_name + "." + op.bucket_id;
+  int ret = cls_cxx_map_remove_key(hctx, key);
+  return ret;
+}
+
+const string resharding_attr = "resharding";
+
+static int rgw_set_bucket_resharding(cls_method_context_t hctx, bufferlist *in,  bufferlist *out)
+{
+  cls_rgw_set_bucket_resharding_op op;
+
+  bufferlist::iterator in_iter = in->begin();
+  try {
+    ::decode(op, in_iter);
+  } catch (buffer::error& err) {
+    CLS_LOG(1, "ERROR: cls_rgw_set_bucket_resharding: failed to decode entry\n");
+    return -EINVAL;
+  }
+
+  bufferlist bl;
+  ::encode(true, bl);
+  int ret = cls_cxx_setxattr(hctx, resharding_attr.c_str(), &bl);
```

Comment:
@oritwas I think it'd be better to have the resharding info as an extra field in the bucket index header. Can use `read_bucket_header()` and `write_bucket_header()` to read header. Modify `struct rgw_bucket_dir_header`. Also, create a new structure for keeping the resharding info (even if it's only a bool).